### DISCUSSION TITLE
Eclipse Sirius

### DISCUSSION THREAD

**Discussion by MarvinLudwig** [eclipse-glsp/glsp#1091] [General] [answered]
How is GLSP different from Sirius?

**Comment by planger** [comment, marked as answer]
Hi @MarvinLudwig,

there are several differences and a few commonalities between GLSP and Sirius.

Let's start with the commonalities:
* Both can be used to develop diagram editors
* In case of Sirius Web (but not Sirius Desktop)
  * Both have a client-server architecture, whereas the model is managed in a backend
  * Both use [Eclipse Sprotty](https://sprotty.org/) as a frontend rendering framework

The most important differences are:

* GLSP is a programming framework for developing diagram editors. Thus, you write e.g. Typescript code to implement certain interfaces expected by the framework for developing your diagram editor and your code interacts directly with GLSP library code on the same level.
* Sirius is built around a declarative model (odesign) describing how an EMF domain model is mapped to a graphical notation model (describing the diagram), how editing tools are configured, etc. This odesign model is interpreted at runtime to construct the diagram editor dynamically, for more complex queries a specific query language (AQL) is used. In a sense it is a modeling editor to create modeling editors.
* GLSP is designed to be very modular for adopters of GLSP. It provides different compatible frameworks and integrations for many aspects of a diagram editor. For instance, this allows choosing between Java or Typescript for implementing the diagram backend -- or even no backend at all, hook up any data source to store and manage your underlying data models (e.g. EMF, JSON, XML, etc.) and integrate with different tool platforms (VS Code, Eclipse Theia, Eclipse RCP, plain web app, etc.) or other types views (property views, other editors, etc.). See also [GLSP's project templates](https://eclipse.dev/glsp/documentation/gettingstarted/).
* Sirius is a one-stop solution with one Java/EMF-based model management server and one Typescript client. Thus, all usual aspects of a modeling tool are coming from this single project, from model definition, management, diagram notation, other views interacting with the model.

In my personal view, Sirius is very fast to get started and also great for rapid prototyping, due to its model-based way of defining diagrams and its one-track architecture. Naturally, for use cases that don't exactly fit this architecture, this approach may entail limitations in flexibility (e.g. using a different model management or aiming at a homogeneous Typescript-based developer experience) or may add complexity as soon as you need customizations that go beyond what can be expressed easily with the declarative model.
On the other hand, GLSP as a programming framework requires code to be written for developing a diagram editor, which may be more "technical" to get started in comparison to "clicking together" a diagram editor. Also you need to select the combination of components that exactly fit your use case (e.g. for using JSON as a data format, Typescript on the client and VS Code as a tool platform). On the upside, this gives you more flexibility (technology/architecture selection and integration with other components) and fine-grained control to realize specific requirements as they can just be implemented, integrated and debugged with common means of writing Typescript code with a Typescript framework (or Java if you choose the Java-based server framework).
